crates/ruma-events/src/room/message/url_preview.rs

```diff
@@ -0,0 +1,501 @@
+    pub fn canonical_url(url: String) -> Self {
+        UrlPreview {
+            matched_url: None,
+            url: Some(url),
+            image: None,
+            description: None,
+            title: None,
+        }
+    }
+
+    /// Whether this preview contains an actual preview or the users homeserver
+    /// should be asked for preview data instead.
+    pub fn contains_preview(&self) -> bool {
+        self.url.is_some()
+            || self.title.is_some()
+            || self.description.is_some()
+            || self.image.is_some()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use assert_matches2::assert_matches;
+    use js_int::uint;
+    use ruma_common::{mxc_uri, serde::Base64};
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+    use super::{super::text::TextMessageEventContent, *};
+    use crate::room::{EncryptedFile, JsonWebKey};
+
+    fn dummy_jwt() -> JsonWebKey {
+        JsonWebKey {
+            kty: "oct".to_owned(),
+            key_ops: vec!["encrypt".to_owned(), "decrypt".to_owned()],
+            alg: "A256CTR".to_owned(),
+            k: Base64::new(vec![0; 64]),
+            ext: true,
+        }
+    }
+
+    fn encrypted_file() -> EncryptedFile {
+        let mut hashes: BTreeMap<String, Base64> = BTreeMap::new();
+        hashes.insert("sha256".to_owned(), Base64::new(vec![1; 10]));
+        EncryptedFile {
+            url: mxc_uri!("mxc://localhost/encryptedfile").to_owned(),
+            key: dummy_jwt(),
+            iv: Base64::new(vec![1; 12]),
+            hashes,
+            v: "v2".to_owned(),
+        }
+    }
+
+    #[test]
+    fn created_preview_image_to_json() {
+        let expected_result = json!({
+            "og:image": "mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO"
+        });
+
+        let preview =
+            PreviewImage::plain(mxc_uri!("mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO").to_owned());
+
+        assert_eq!(to_json_value(&preview).unwrap(), expected_result);
+
+        let encrypted_result = json!({
+            "beeper:image:encryption": {
+                "hashes" : {
+                    "sha256": "AQEBAQEBAQEBAQ",
+                },
+                "iv": "AQEBAQEBAQEBAQEB",
+                "key": {
+                    "alg": "A256CTR",
+                    "ext": true,
+                    "k": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+                    "key_ops": [
+                        "encrypt",
+                        "decrypt"
+                    ],
+                    "kty": "oct",
+                },
+                "v": "v2",
+                "url": "mxc://localhost/encryptedfile",
+            },
+        });
+
+        let preview = PreviewImage::encrypted(encrypted_file());
+
+        assert_eq!(to_json_value(&preview).unwrap(), encrypted_result);
+    }
+
+    #[test]
+    fn parsing_regular_example() {
```

Comment:
nit: here and below, we usually use `deserialize` instead of `parse` in test names.